nengo/params.py

```diff
@@ -130,8 +132,29 @@ def get_default(self, obj):
     def set_default(self, obj, value):
         if not self.configurable:
             raise ConfigError("Parameter '%s' is not configurable" % self)
-        self.validate(obj, value)
-        self._defaults[obj] = value
+        self._defaults[obj] = self.coerce(obj, value)
+
+    def check_type(self, instance, value, type_):
+        if value is not None and not isinstance(value, type_):
+            if isinstance(type_, tuple):
+                type_str = " or ".join((t.__name__ for t in type_))
+            else:
+                type_str = type_.__name__
+            raise ValidationError("Must be of type %r (got type %r)."
+                                  % (type_str, type(value).__name__),
+                                  attr=self.name, obj=instance)
+
+    def coerce(self, instance, value):
+        if isinstance(value, DefaultType):
+            raise ValidationError("Default is not a valid value. To reset a "
+                                  "parameter, use 'del'.",
+                                  attr=self.name, obj=instance)
+        if self.readonly and instance in self.data:
```

Comment:
Yeah, these aren't mutually exclusive, though it's true that the same thing will happen if they're `elifs`.

For me, the reason I prefer `if` statements here is because each check is an independent thing. Typically with `if`/`elif` chains, you have to keep in mind the previous conditions when looking at each condition. E.g.,

```python
if a:
    # do A things
elif b:
    # do B things
elif c:
    # do C things
```

The second condition is actually entered when `not a and b`; similarly the third condition is `not a and not b and c`. The ability to simplify the condition with `elif`s is convenient, and usually the logic is pretty clear. But in this case, it doesn't help the reader to have the `elif` because the previous if statements are irrelevant to the current one. Like, it sort of matters in that they will never be evaluated if the previous if statement is true, but you don't have to think about the previous ones when looking at an individual if statement.

Another reason why I prefer `if`s here is that it follows a typical and useful pattern:

```python
if thing.will_fail:
    # quickly exit this function because we're going to fail; either
    return None  # keep going because failure is ok
    raise Error  # stop because we can't continue
```

When I see that kind of thing, I know what's up without having to think much about it. In this case, most of `coerce` is a bunch of instances of this pattern, so I find it easier to scan than if they had been elifs. Either would act the same, though.